### DISCUSSION TITLE
segmentedTimeLine.addException

### DISCUSSION THREAD

**Discussion by marksmeets** [jfree/jfreechart#299] [Q&A] [answered]
I think I don't understand SegmentedTimeLine.addException. I'm using it to draw bar charts on intraday stock data. Let's assume the bars have size 15 minutes. Data typically comes in between 09:00 and 17:30 but this may vary per day and per stock. I want a continuous display of the bars, without gaps when there is no data. It seems the way to deal with this is by applying a segmentedTimeLine on the domain axis. My idea is to whenever a new bar needs to be added, check if there is a gap between the previous bar and the current (new to be added) bar. Like this:

```    
if ( t.getFirstMillisecond() - (_ohlcSeries.getPeriod(_ohlcSeries.getItemCount()-1).getFirstMillisecond() + getTimeFrame().barLengthMs() ) > 0)
{
        _log.info("Gap between " + t + " and " + _ohlcSeries.getPeriod(_ohlcSeries.getItemCount()-1));
        _mondayFridayTimeline.addException(_ohlcSeries.getPeriod(_ohlcSeries.getItemCount()-1).getFirstMillisecond() + getTimeFrame().barLengthMs()  + 1, t.getFirstMillisecond()-1);
}
```

Here 't' is the RegularTimePeriod of the new bar and getTimeFrame().barLengthMS() returns the number of milliseconds in bar. The detection of gaps seems to work. Then I add an exception to the TimeLine already present (called _mondayFridayTimeline), but this doesn't seem to have any effect; I see gaps everywhere. 

Am I understanding the purpose of addException incorrectly? Or is there a better way to handle this. 'Guessing' beforehand what the gaps will be and building repeating segments beforehand doesn't seem like a solution I could use.

Thanks,
Mark.

**Comment by trashgod** [comment, marked as answer]
As outlined [here](https://github.com/jfree/jfreechart#migration-from-jfreechart-10x), "the `SegmentedTimeline` class has been removed…" As suggested [here](https://stackoverflow.com/a/55312264/230513), you can implement `Timeline` as needed, but I'm not familiar with how `addException()` worked.

**Reply by marksmeets** [reply]
Thanks, your suggestions look very promising and may be a better fit. In the meantime I think I've figured out what was wrong: the _mondayFridayTimeline has segment size equal to a day, while I'm adding exceptions of length less than a day. Once I defined _mondayFridayTimeline with segment size equal to a minute the gaps disappeared. It does give some rendering issues though, like date labels on top of eachother. I dont' think I'll be able to fix that (way over my head). For that reason I think I'll switch to the approach pointed out in your suggestion: use an XYDataset and make the domain axis have labels based upon dates associated with the 'x' in the XYSeries.

**Reply by matoos32** [reply]
Hi @marksmeets you may also be interested in https://github.com/jfree/jfreechart/issues/197#issuecomment-1154674436

**Reply by marksmeets** [reply]
Thanks Matt, that looks awesome!
